truss-build/src/truss/build/image.py

```diff
@@ -0,0 +1,165 @@
+from .commands import *
+from typing import Dict, List, Optional, Union
+from pathlib import Path
+from .utils import flatten_str_args, make_pip_install_args
+import shlex
+
+# TODO(bola): support secrets
+
+
+class Image:
+    def __init__(self) -> None:
+        self._base_image = ""
+        self._commands: list[Command] = []
+        self._base_image: Optional[str] = None
+
+    # TODO(support static constructors for different ways of pulling base iamge)
+    # Image.from_dockerfile, from_aws_ecr, from_gcp_artifact_registry, from_registry
+    # also can add common base images: Image.truss_base()....
+
+    def serialize(self) -> str:
+        dockerfile = "# syntax = docker/dockerfile:1\n"  # Support BuildKit
+        dockerfile += "# Auto-generated by truss.build, do not edit!\n"
+        for command in self._commands:
+            dockerfile += command.serialize() + "\n"
+        return dockerfile
+
+    def apt_install(
+        self,
+        *packages: Union[str, List[str]],
+    ) -> "Image":
+        """Install a list of Debian packages using `apt`.
+
+        **Example**
+
+        ```python
+        image = Image.debian_slim().apt_install("git")
+        ```
+        """
+        pkgs = flatten_str_args("apt_install", "packages", packages)
+        if not pkgs:
+            return self
+        package_args = " ".join(shlex.quote(pkg) for pkg in pkgs)
+        self._commands.extend(
+            [
+                RunCommand("apt-get update"),
+                RunCommand(f"RUN apt-get install -y {package_args}"),
+            ]
+        )
+        return self
+
+    def run_commands(
+        self,
+        *commands: Union[str, List[str]],
+    ) -> "Image":
+        """Extend an image with a list of shell commands to run."""
+        cmds = flatten_str_args("run_commands", "commands", commands)
+        if not cmds:
+            return self
+        self._commands.extend([RunCommand(cmd) for cmd in cmds])
+        return self
+
+    def workdir(self, path: str) -> "Image":
+        """Sets the working directory for subequent image build steps.
+
+        :param path: A path to set the workdir.
+
+        **Example**
+
+        ```python
+        image = (
+            Image.debian_slim()
+                .run_commands("git clone https://xyz app")
+                .workdir("/app")
+                .run_commands("yarn install")
+        )
+        ```
+        """
+        self._commands.append(WorkdirCommand(path))
+        return self
+
+    def env(self, vars: Dict[str, str]) -> "Image":
```

Comment:
Nit: to denote that `vars` is immutable ("const") for this function, use `from typing import Mapping`.